From 9a9b4c59babc60e15a79cf7db8167d42e68e44b9 Mon Sep 17 00:00:00 2001
From: Jim Meyering <meyering@fb.com>
Date: Sat, 02 Nov 2013 19:15:48 +0000
Subject: grep: fix regression with -P vs. invalid UTF-8 input

* src/pcresearch.c (Pexecute): Don't abort upon unexpected
PCRE-specific error code.  Explicitly handle PCRE_ERROR_BADUTF8,
and change the default to print a diagnostic including the unhandled
integer PCRE error code and exit with status 2.
* tests/pcre-invalid-utf8-input: New file.
* tests/Makefile.am (TESTS): Add it.
* NEWS (Bug fixes): Mention it.
* THANKS: Update.
Reported by Dave Reisner in http://bugs.gnu.org/15758.
---
--- a/NEWS
+++ b/NEWS
@@ -44,6 +44,12 @@
   For example, \s*, \s+, \s? and \s{3} would all malfunction in a
   multi-byte locale.  [bug introduced in grep-2.15]
 
+  The fix to make grep -P work better with UTF-8 made it possible for
+  grep to evoke a larger set of PCRE errors, some of which could trigger
+  an abort.  E.g., this would abort:
+    printf '\x82'|LC_ALL=en_US.UTF-8 grep -P y
+  Now grep handles arbitrary PCRE errors.  [bug introduced in grep-2.15]
+
 
 * Noteworthy changes in release 2.14 (2012-08-20) [stable]
 
--- a/THANKS
+++ b/THANKS
@@ -20,6 +20,7 @@
 Corinna Vinschen           <vinschen@redhat.com>
 Dagobert Michelsen         <dam@opencsw.org>
 Daisuke GOTO               <gotoh@m-design.com>
+Dave Reisner               <d@falconindy.com>
 David Clissold             <cliss@austin.ibm.com>
 David J MacKenzie          <djm@catapult.va.pubnix.com>
 David O'Brien              <obrien@freebsd.org>
--- a/src/pcresearch.c
+++ b/src/pcresearch.c
@@ -256,9 +256,20 @@
           error (EXIT_TROUBLE, 0,
                  _("exceeded PCRE's backtracking limit"));
 
+        case PCRE_ERROR_BADUTF8:
+          error (EXIT_TROUBLE, 0,
+                 _("invalid UTF-8 byte sequence in input"));
+
         default:
-          abort ();
+          /* For now, we lump all remaining PCRE failures into this basket.
+             If anyone cares to provide sample grep usage that can trigger
+             particular PCRE errors, we can add to the list (above) of more
+             detailed diagnostics.  */
+          error (EXIT_TROUBLE, 0, _("internal PCRE error: %d"), e);
         }
+
+      /* NOTREACHED */
+      return -1;
     }
   else
     {
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -79,6 +79,7 @@
   options					\
   pcre						\
   pcre-abort					\
+  pcre-invalid-utf8-input			\
   pcre-utf8					\
   pcre-z					\
   prefix-of-multibyte				\
--- /dev/null
+++ b/tests/pcre-invalid-utf8-input
@@ -0,0 +1,25 @@
+#! /bin/sh
+# Ensure that grep -P doesn't abort for invalid multi-byte input
+#
+# Copyright (C) 2013 Free Software Foundation, Inc.
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.
+
+. "${srcdir=.}/init.sh"; path_prepend_ ../src
+require_pcre_
+require_en_utf8_locale_
+
+fail=0
+
+printf '\202\n' > in || framework_failure_
+printf 'grep: invalid UTF-8 byte sequence in input\n' \
+   > exp-err || framework_failure_
+
+LC_ALL=en_US.UTF-8 grep -P anything in > out 2> err
+test $? = 2 || fail=1
+compare /dev/null out || fail=1
+compare exp-err err || fail=1
+
+Exit $fail
